src/engagement_db_coda_sync/lib.py

```diff
@@ -67,6 +68,72 @@ def _add_message_to_coda(coda, coda_dataset_config, ws_correct_dataset_code_sche
     coda.add_message_to_dataset(coda_dataset_config.coda_dataset_id, coda_message)
 
 
+def _code_for_label(label, code_schemes):
+    """
+    Returns the code for the given label.
+
+    Handles duplicated scheme ids (i.e. schemes ending in '-1', '-2' etc.).
+    Raises a ValueError if the label isn't for any of the given code schemes.
+
+    :param label: Label to get the code for.
+    :type label: core_data_modules.data_models.Label
+    :param code_schemes: Code schemes to check for the given label.
+    :type code_schemes: list of core_data_modules.data_models.CodeScheme
+    :return: Code for the label.
+    :rtype: core_data_modules.data_models.Code
+    """
+    for code_scheme in code_schemes:
+        if label.scheme_id.startswith(code_scheme.scheme_id):
+            return code_scheme.get_code_with_code_id(label.code_id)
+
+    raise ValueError(f"Label's scheme id '{label.scheme_id}' is not in any of the given `code_schemes` "
+                     f"(these have ids {[scheme.id for scheme in code_schemes]})")
+
+
+def _get_ws_code(coda_message, coda_dataset_config, ws_correct_dataset_code_scheme):
+    """
+    Gets the WS code assigned to a Coda message, if it exists, otherwise returns None.
+
+    :param coda_message: Coda message to check for a WS code.
+    :type coda_message: core_data_modules.data_models.Message
+    :param coda_dataset_config: Dataset configuration to use to interpret this message's labels.
+    :type coda_dataset_config: src.engagement_db_coda_sync.configuration.CodaDatasetConfiguration
+    :param ws_correct_dataset_code_scheme: WS - Correct Dataset code scheme.
+    :type ws_correct_dataset_code_scheme: core_data_modules.data_models.CodeScheme
+    :return: WS code assigned to this message, if it exists.
+    :rtype: core_data_modules.data_models.Code | None
+    """
+    normal_code_schemes = [c.code_scheme for c in coda_dataset_config.code_scheme_configurations]
+    ws_code_scheme = ws_correct_dataset_code_scheme
+
+    # Check for a WS code in any of the normal code schemes
+    ws_code_in_normal_scheme = False
+    for label in coda_message.get_latest_labels():
+        if label.scheme_id == ws_code_scheme.scheme_id:
+            continue
+        code = _code_for_label(label, normal_code_schemes)
+        if code.control_code == Codes.WRONG_SCHEME:
+            ws_code_in_normal_scheme = True
+
+    # Check for a code in the WS code scheme
+    code_in_ws_scheme = False
+    ws_code = None
+    for label in coda_message.get_latest_labels():
+        if label.scheme_id == ws_code_scheme.scheme_id:
+            code_in_ws_scheme = True
+            ws_code = ws_code_scheme.get_code_with_code_id(label.code_id)
```

Comment:
Can this code block come before the `continue` statement in line 113?

Reply:
It could, but I kept it separate because we're searching for two different things in each loop, and I was finding the logic was getting a bit tough to read if we did it all in one. What do you think?

Reply:
yeah, I get you, I think then it will be better if we update the condition on L112 to check for inequality & remove the continue statement.

Reply:
agreed, will make that change instead, thanks